contracts-available/FlightDelayStaking.sol

```diff
@@ -32,230 +48,245 @@ contract FlightDelayRiskPool is Ownable {
+        collatFactor = _collatFactor;
     }
 
     /**
-     * Calculate the required DIP tokens to stake the requested amount of stable asset
-     * Currently we use a fixed factor e.g. 1.0
-     *
-     * @param _stableStake
-     * @returns _requiredStake Amount of DIP tokens required.
+     * @dev Stake stable and dip tokens
+     * @param _stake is the amount of dip token
      *
      */
     function stake(uint256 _stake) public payable {
-
-        stake currentStakersStake = stakeBalances[msg.sender];
-        increaseCurrentStake(_stake);
-        uint256 totalStableStake = currentSingleStake.stable + msg.value;
-        uint256 totalDipStake = currentStake.dip + _stake;
+        Stake memory currentStakersStake = stakeBalances[msg.sender];
+        uint256 totalStableStake = currentStakersStake.stable + msg.value;
+        uint256 totalDipStake = currentStakersStake.dip + _stake;
         uint256 requiredDip = calculateRequiredDip(totalStableStake);
-        require (totalDipStake <= requiredDip, "Stake to high");
-        require (DipTokenContract.transferFrom(msg.sender, this, stake), "DIP could not be staked");
-        stakeBalances[msg.sender] = stake(totalStableStake, totalDipStake);
+        require(totalDipStake <= requiredDip, "Stake to high");
+        require(
+            dipTokenContract.transferFrom(msg.sender, address(this), _stake),
+            "DIP could not be staked"
+        );
+        stakeBalances[msg.sender] = Stake(totalStableStake, totalDipStake);
+        currentStake += msg.value;
 
-        emit LogDIPStaked(stake, msg.value, totalDipStake, totalStableStake);
+        emit LogDIPStaked(_stake, msg.value, totalDipStake, totalStableStake);
+    }
 
+    /**
+     * @dev calculates required dip token amount from stable token amount. _requiredStake Amount of DIP tokens required.
+     * @param _stableStake is the amount of stable coin
+     */
+    function calculateRequiredDip(uint256 _stableStake)
+        public
+        view
+        returns (uint144 _requiredStake)
+    {
+        _requiredStake = FixedPoint
+            .fraction(stakingRelation.val, stakingRelation.div)
+            .mul(_stableStake)
+            .decode144();
     }
 
     /**
-     * Calculate the required DIP tokens to stake the requested amount of stable asset
-     * Currently we use a fixed factor e.g. 1.0
+     * @dev locks the capacity
+     * @param _capacity is the amount to lock
      *
-     * @param _stableStake
-     * @returns _requiredStake Amount of DIP tokens required.
+     */
+    function lockCapacity(uint256 _capacity) public {}
+
+    /**
+     * @dev unlocks the capacity
+     * @param _capacity is the amount to unlock
      *
      */
-    function calculateRequiredDip(uint256 _stableStake) public view returns (uint256 _requiredStake) {
-        _requiredStake = fixedMul(_stableStake, stakingRelation);
+    function unlockCapacity(uint256 _capacity) public {}
```

Comment:
here we first adjust the lockedStake value.
Then we can process the available unstakeRequests and see if one / some of them can be fulfilled.